src/librustc_resolve/diagnostics.rs

```diff
@@ -397,6 +397,66 @@ impl Bar {
 ```
 "##,
 
+E0417: r##"
+A static variable was referenced in a pattern. Example of erroneous code:
+
+```
+static FOO : i32 = 0;
+
+match 0 {
+    FOO => {} // error: static variables cannot be referenced in a
+              //        pattern, use a `const` instead
+    _ => {}
+}
+```
+
+The compiler needs to know the value of the pattern at compile time,
+which is not possible with a `static` variable. Please verify that the
```

Comment:
nit: saying "is not possible" may be re-opening a can of worms by implicitly inviting back those who were arguing, back when we first split up the two forms, that the `static` / `const` distinction was unnecessary.

I would prefer something like:

> The compiler needs to know the value of the pattern at compile time; compile-time patterns can defined via `const` or `enum` items. Please verify that the identifier is spelled correctly, and if so, use a `const` instead of `static` to define it.
> Example:
> ...